package.json

```diff
@@ -39,10 +39,11 @@
   "license": "MIT",
   "dependencies": {
     "axios": "0.27.2",
+    "cli-progress": "^3.11.2",
```

Comment:
One final request before merging, can you please move this to devDependencies since it's only used in the test script.

Reply:
Hello **@kwypchlo,**

**Sorry for the long answer and my concerns, but I'm not only a developer but also an end customer.**

I think it wouldn't make sense, since "cli-progress" belongs to the "onUploadProgress" function, and prints a clean one-line progressbar in the console output.
I implemented this code instead of https://sdk.skynetlabs.com/?javascript--browser#setting-additional-options:

**Code for "skynet-js":**
```
import { SkynetClient } from "skynet-js";

// Set an upload progress tracker.
const onUploadProgress = (progress, { loaded, total }) => {
  console.info(`Progress ${Math.round(progress * 100)}%`);
};

// Initialize the client.
const client = new SkynetClient("https://siasky.net", { onUploadProgress });

async function uploadExample() {
  try {
    const skylink = await client.uploadFile(file);
  } catch (error) {
    console.log(error);
  }
}
```


**Code for "nodejs-skynet":**
```
import { SkynetClient, onUploadProgress } from "nodejs-skynet";

// Initialize the client.
const client = new SkynetClient("https://siasky.net", { onUploadProgress });

async function uploadExample() {
  try {
    const skylink = await client.uploadFile(file);
  } catch (error) {
    console.log(error);
  }
}
```


Because the code of "skynet-js" in the browser console prints every percentage step in a new line, which is also ok for the browser.

But for "nodejs-skynet" in the OS console it doesn't look anything if every percentage step is written on a new line, so I integrated a ProgressBar for nodejs that outputs everything in one line.

I think if I we shouldn't put "cli-progress" in "devDependencies" as it belongs to the code for "onUploadProgress" and I'm not sure if it still works after that when published with "--production" flag.
Without "cli-progress" "onUploadProgress" would no longer output anything in the console, which would confuse the end customer.

**The "devDependencies" only get installed when I install the source code from the github repo, but not when I install "SkynetLabs/nodejs-skynet" from a cdn with npm install?
Am I right or wrong?**

I find my integration in "nodejs-skynet" to be more user friendly as the end customer just has to integrate the code to make onUploadProgress work:
```
import { SkynetClient, onUploadProgress } from "nodejs-skynet";
const client = new SkynetClient("https://siasky.net", { onUploadProgress });
```

In the "skynet-js" browser, the end customer must ensure that he also creates the "onUploadProgress" himself or copies it from the skd-doc example.
I personally also had the problem getting "onUploadProgress" to work in the beginning, until I saw in the sdk-doc that I also need this code:
```
// Set an upload progress tracker.
const onUploadProgress = (progress, { loaded, total }) => {
  console.info(`Progress ${Math.round(progress * 100)}%`);
};
```

I think the code above is technically seen, also necessary for the function "onUploadProgress" to work in the browser, and in my opinion it should also be integrated into "skynet-js" similar to how I integrated it into "nodejs-skynet", so that it is easier for the end customer to get "onUploadProgress" up and running.

The end customer can then decide for himself whether he uses the default "onUploadProgress" or creates his own so that it works the way he wants.

Which do you think is better?

Reply:
Hello @kwypchlo,

**Short answer:**
It must remain as it is, otherwise "onUploadProgress" for tus in "nodejs-skynet" will no longer work, which is urgently needed for big uploads and also for "parallelUploads".

If I am to change it, it is no longer useful for the end customer, which should actually result in a friendlier use for the end customer.

**How does it look now with the merge, or should I change it???**

I would rather not finish here, because it works and has already taken long enough.
So that I can finally continue to push things (see iusses) in new RP.

Reply:
![Screenshot 2022-09-28 at 10 40 35](https://user-images.githubusercontent.com/3755029/192732274-a03949f4-bb04-4b4f-919c-5fe9a59f7671.png)

This package [cli-progress](https://www.npmjs.com/package/cli-progress) is used in testing script in your codebase. I just noticed that you are importing custom `onUploadProgress` and exporting it as an api of the sdk. This is not something we want to be honest. It's not a good practice to export from test scripts and also that implementation of function `onUploadProgress` could be an example in README so people can copy it (or use a different one if they want) but not something we would want to include in the library itself. The testing script will work fine if we move that package to devDependencies once we remove the `onUploadProgress` export.

As for your end user experience, nodejs-skynet library would not rely on [cli-progress](https://www.npmjs.com/package/cli-progress) in any way, the `onUploadProgress` callback is defined client side and end user is free to use any library to manage progress just like you did in utils_testing.js file. End user will need to add [cli-progress](https://www.npmjs.com/package/cli-progress) package to the project that uses it. Additionally, it would be a bad practice to use it (require it) in a project without installing it directly via package.json because that's an indirect dependency.

So to sum it up, when a client script is using nodejs-skynet it would need to defined the `onUploadProgress` manually and there is nothing preventing using [cli-progress](https://www.npmjs.com/package/cli-progress) in there. Example:

```js
import { SkynetClient } from "nodejs-skynet";
import cliProgress from "cli-progress";

// Set the custom upload progress tracker.
const onUploadProgress = (progress, { loaded, total }) => {
  let progressOutput = Math.floor((loaded * 100) / total);

  if (progressOutput === 0 && fileSizeCounter === 0) {
    process.stdout.write(" The uploading File size is " + total + " bytes.\n\n");
    process.stdout.moveCursor(0, -1);
    fileSizeCounter++;
  }

  // create new progress bar with custom token "speed"
  const bar = new cliProgress.Bar({
    format: "-> uploading [{bar}] " + progressOutput + "% {eta_formatted} ",
    fps: 1,
  });

  // initialize the bar - set payload token "speed" with the default value "N/A"
  bar.start(100, 0, {
    speed: "N/A",
  });
  bar.updateETA(Buffer);
  // update bar value. set custom token "speed" to 125
  bar.update(progressOutput, {
    speed: "122",
  });

  // stop the bar
  bar.stop();

  if (progressOutput === 100) {
    process.stdout.write(`\n`);
    process.stdout.moveCursor(0, -1);
    process.stdout.write(`\n`);
    endProgressCounter++;
    if (endProgressCounter === 2) {
      process.stdout.write(`\n\n`);
    }
  }
  process.stdout.moveCursor(0, -1);
};

const client = new SkynetClient("https://siasky.net", { onUploadProgress });
```

This example above has the same end result as your example (below) but without pushing a specific dependency on all the library users.

```js
import { SkynetClient, onUploadProgress } from "nodejs-skynet";
const client = new SkynetClient("https://siasky.net", { onUploadProgress });
```

Let me know if that explains it in enough details :)

Reply:
Hello @kwypchlo,

What you write above is all clear to me.

However, in order for the "onUploadProgress" function to work, your sdk doc should explain in bold red that the user can either copy the code from the example or write their own code for the "onUploadProgress" output.

Without the additional code, it doesn't work in the browser either, and that's not clearly explained in the sdk-doc.
But this information has already been sent to the team.

What I don't understand is, if everything is fine for you so far and you only want "cli-progress" in "devDependencies" in the "package.json", why didn't you just change one place in the code and then everything merged into the master.

They don't like the way I've integrated it, so what more can I say.